### ISSUE TITLE
Missing proper libstdc++ version on the rescue image

### ISSUE THREAD

**Issue by juliandroid** [Drive-Trust-Alliance/sedutil#21]
When I run **sedutil-cli** on the Rescue image I get the following error:

`sedutil-cli: /usr/lib/libstdc++.so.6: version 'GLIBCXX_3.4.21' not found (required by sedutil-cli)`

The issue again is on my Arch linux machine, but eventually it will bite you later. The problem is that we are building the **sedutil-cli** by using system gcc, while at the same time we are relying on the TinyCore (built with different gcc).

In my case, obviously libstdc++ is newer and the 'GLIBCXX_3.4.21' is missing on the syslinux image.

My suggestion is to build the sedutil-cli statically. The executable will become 2.4M.

```
diff --git linux/CLI/nbproject/Makefile-Release_i686.mk linux/CLI/nbproject/Makefile-Release_i686.mk
index a93b927..eabc01b 100644
--- linux/CLI/nbproject/Makefile-Release_i686.mk
+++ linux/CLI/nbproject/Makefile-Release_i686.mk
@@ -73,7 +73,7 @@ FFLAGS=-Wall
 ASFLAGS=

 # Link Libraries and Options
-LDLIBSOPTIONS=
+LDLIBSOPTIONS=-static

 # Build Targets
 .build-conf: ${BUILD_SUBPROJECTS}
```

Doing so, the program now runs correctly.

The same patch is required for the UEFI64 image:

```
diff --git linux/CLI/nbproject/Makefile-Release_x86_64.mk linux/CLI/nbproject/Makefile-Release_x86_64.mk
index b9ce233..3d0a679 100644
--- linux/CLI/nbproject/Makefile-Release_x86_64.mk
+++ linux/CLI/nbproject/Makefile-Release_x86_64.mk
@@ -73,7 +73,7 @@ FFLAGS=-Wall
 ASFLAGS=

 # Link Libraries and Options
-LDLIBSOPTIONS=
+LDLIBSOPTIONS=-static

 # Build Targets
 .build-conf: ${BUILD_SUBPROJECTS}
```

**Comment by r0m30** [comment]
Thanks.  I'm considering using buildroot to replace tinycore for the rescue system, right now I'm not exploiting it to it's fullest.  I think that if I created a board or packages for the PBA's and sedutil I could do a complete and consistent build to avoid these types of issues.  

Modifying the makefiles directly will not work, Netbeans will overwrite them on a whim.  I'll have to modify the project properties in netbeans.

**Comment by juliandroid** [comment]
Sure. I compile on the command line and that's why I'm patching the makefiles :)

**Comment by juliandroid** [comment]
I'm facing similar problem, this time while booting from the shadow UEFI64 PBA.

> /sbin/linuxpba: error while loading shared libraries: libncursesw.so.6: cannot open shared object file: No such file or directory

The first problem is that LinuxPBA depends on ncurses (-lcurses while linking) which probably is different version that the one available in syslinux.
The second problem is that my Arch Linux packages doesn't include statically linked ncurses, so I can't even statically link as I did for sedutil-cli. That makes building process for me blocker :(

**Comment by konstmonst** [comment]
Here is a simplifed linux PBA with ncurses removed, which can be compiled staticly. 
This one also properly enumerates avalible block devices. I am using getpass() so you can remove everything from GetPassPhrase.cpp. I refactored UnlockSEDs a bit so I can read it better.
Here is the code, feel free to use it under the License of this project.

main.cpp:
`
using namespace std;
#include <unistd.h>
#include <sys/reboot.h>
#include "log.h"
#include "UnlockSEDs.h"
#include <unistd.h>

int main(int argc, char** argv)
{
	char* pw = getpass("Please enter pass-phrase to unlock OPAL drives: ");
	if (pw)
	{
		if (UnlockSEDs(pw) == 0)
		{
			sync();
			reboot(RB_AUTOBOOT);
			return EXIT_SUCCESS;
		}
	}
	return EXIT_FAILURE;
}
`
UnlockSED.cpp:
`
#include "UnlockSEDs.h"
#include "DtaDevGeneric.h"
#include "DtaDevOpal1.h"
#include "DtaDevOpal2.h"
#include <unistd.h>
#include <dirent.h>
#include <stdio.h>

static int UnlockDevice(DtaDev& d, const char* devref, char* password)
{
	d.no_hash_passwords = false;

	printf("Unlocking %s\n", devref);
	int result = -1;
	if (d.MBREnabled())
	{
		if (d.setMBRDone(1, password) == 0)
		{
			if (d.setLockingRange(0, OPAL_LOCKINGSTATE::READWRITE, password) == 0)
			{
				result = 0;
			}
			else
			{
				LOG(E) << "Unlock failed - unable to set LockingRange 0 RW";
			}
		}
		else
		{
			LOG(E) << "Unlock failed - unable to set MBRDone";
		}
	}
	else
	{
		LOG(I) << "MBR not enabled";
		result = 0;
	}

	return result;
}

uint8_t UnlockSEDs(char * password)
{
	int result = -1;

	DIR* d;
	struct dirent* dir;
	d = opendir("/sys/block/");
	if (d)
	{
		LOG(D4) << "Enter UnlockSEDs";
		printf("Scanning...\n");

		while ((dir = readdir(d)) != NULL)
		{
			char devref[16];
			if (dir->d_name[0] != '.')
			{
				snprintf(devref, sizeof(devref), "/dev/%s", dir->d_name);
				DtaDevGeneric tempDev(devref);
				if (tempDev.isPresent())
				{
					if (tempDev.isOpal2())
					{
						DtaDevOpal2 d(devref);
						result = UnlockDevice(d, devref, password);
					}
					else if (tempDev.isOpal1())
					{
						DtaDevOpal1 d(devref);
						result =  UnlockDevice(d, devref, password);
					}
					else
					{
						printf("Drive %s not supported\n", devref);
					}
				}
			}
		}
	}

	closedir(d);
	return result;
}
`

**Comment by juliandroid** [comment]
Thanks for sharing.
My current workaround is to compile the ncurses aside of the sedutil project and later modify the makefile to use the generated ncurses.a (object code archive file for static linking). That's way I have LinuxPBA and sedutil-cli statically linked, BUT that is just **workaround**, nothing more.

**Comment by r0m30** [comment]
This should be resolved with V1.15
